servers/visual/visual_server_scene.cpp

```diff
@@ -3035,6 +3042,47 @@ void VisualServerScene::_prepare_scene(const Transform p_cam_transform, const Ca
 
 			InstanceGeometryData *geom = static_cast<InstanceGeometryData *>(ins->base_data);
 
+			// Calculate instance->depth from the camera.
+			const Vector3 aabb_center = ins->transformed_aabb.position + (ins->transformed_aabb.size * 0.5);
+			ins->depth = near_plane.distance_to(aabb_center);
+			ins->depth_layer = CLAMP(int(ins->depth * 16 / z_far), 0, 15);
```

Comment:
This calculation seems only required here if LOD is active for this instance. This could be set efficiently with e.g. a bool when setting the LOD params for the instance. i.e.:
```
if (ins->lod_active) ... do lod calcs
```
In addition the depth calculated here could potentially be reused in many cases later in the line 3305 part, to save doing the same calculation twice. Although that would need some perf testing to check it is a win.

Reply:
> This calculation seems only required here if LOD is active for this instance. This could be set efficiently with e.g. a bool when setting the LOD params for the instance. i.e.:

Do we need a dedicated boolean field for this? We can check if LOD is enabled by checking whether the min range *or* max range is greater than 0 (exclusive). Surely two float comparisons are cheap enough to perform :slightly_smiling_face:

Reply:
I haven't profiled this, but imo a good aim is to make any new code as cheap as possible for instances that aren't using LOD. We essentially should aim not to be doing extra unnecessary calculations for the 99% of times where people aren't using LOD on an instance.

Doing unneeded distance calculations, and checking two float values against `CMP_EPSILON` is not as cheap as just detecting a bool is set to false which will likely be the cheapest conditional (bear in mind on some mobile floating point math could even be emulated). This double float check is additionally now not required at runtime _should the bool be set to true_.

There are lots of occasions where just doing things unoptimized are fine (where code clarity is important, and it isn't performance sensitive), but inside the renderer imo we should strive to make it as efficient as possible as there is a fair amount of churn, and we are just wasting cycles and battery on mobile.

Looking at the `Instance : RasterizerScene::InstanceBase` structure, it looks like `RID lod_instance` is unused so could be removed. You could for instance add a bool up with the `update_materials` bools, and you could possibly make them 1 bit too (e.g. `bool update_aabb : 1;`), so they would likely be no more than 32 bit between them.

In `instance_geometry_set_draw_range()` you could just do something like this:
```
instance->using_lod = (instance->lod_begin > CMP_EPSILON) || (instance->lod_end > CMP_EPSILON);
```
Therefore the float check is done once (rarely) when setting the params, and the cheap bool test can be done at runtime on each frame on each `prepare_scene`.